File: frontend/src/lib/offHours.ts
export interface OffHoursWindow {
  days?: string[]
  start: string
  end: string
}

export interface OffHoursConfig {
  enabled: boolean
  blocked: boolean
  unblockAt?: string
  message?: string
  windows?: OffHoursWindow[]
}

let cachedConfig: OffHoursConfig | null = null
let configFetchPromise: Promise<OffHoursConfig> | null = null

const DEFAULT_MESSAGE = 'Time to rest!'

export const fetchOffHoursConfig = async (): Promise<OffHoursConfig> => {
  if (cachedConfig !== null) {
    return cachedConfig
  }

  if (configFetchPromise !== null) {
    return configFetchPromise
  }

  configFetchPromise = fetch('/api/v1/off_hours')
    .then(async (response) => {
      if (!response.ok) {
        return { enabled: false, blocked: false }
      }
      const data = await response.json()
      cachedConfig = data
      return data
    })
    .catch(() => {
      return { enabled: false, blocked: false }
    })
    .finally(() => {
      configFetchPromise = null
    })

  return configFetchPromise
}

export const clearOffHoursCache = (): void => {
  cachedConfig = null
}

const parseTimeOfDay = (s: string): { hour: number; minute: number } | null => {
  const parts = s.split(':')
  if (parts.length !== 2) return null

  const hour = parseInt(parts[0], 10)
  const minute = parseInt(parts[1], 10)

  if (isNaN(hour) || isNaN(minute)) return null
  if (hour < 0 || hour > 23) return null
  if (minute < 0 || minute > 59) return null

  return { hour, minute }
}

const timeOfDayMinutes = (hour: number, minute: number): number => {
  return hour * 60 + minute
}

const dayNames = ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday']

const containsDay = (days: string[] | undefined, day: string): boolean => {
  if (!days || days.length === 0) return true
  return days.some((d) => d.toLowerCase() === day.toLowerCase())
}

export const isTimeInWindow = (
  t: Date,
  window: OffHoursWindow
): { blocked: boolean; unblockAt: Date | null } => {
  const startParsed = parseTimeOfDay(window.start)
  const endParsed = parseTimeOfDay(window.end)

  if (!startParsed || !endParsed) {
    return { blocked: false, unblockAt: null }
  }

  const dayName = dayNames[t.getDay()]
  const currentMinutes = timeOfDayMinutes(t.getHours(), t.getMinutes())
  const startMinutes = timeOfDayMinutes(startParsed.hour, startParsed.minute)
  const endMinutes = timeOfDayMinutes(endParsed.hour, endParsed.minute)

  const crossesMidnight = endMinutes < startMinutes

  if (crossesMidnight) {
    // Window crosses midnight, e.g., 23:00 to 07:00
    // Check if we're in the "before midnight" part (on start day)
    // or the "after midnight" part (on end day, which is start day + 1)
    const previousDayName = dayNames[(t.getDay() + 6) % 7]

    if (containsDay(window.days, dayName) && currentMinutes >= startMinutes) {
      // We're in the before-midnight portion on a matching day
      const unblockAt = new Date(t)
      unblockAt.setDate(unblockAt.getDate() + 1)
      unblockAt.setHours(endParsed.hour, endParsed.minute, 0, 0)
      return { blocked: true, unblockAt }
    }

    if (containsDay(window.days, previousDayName) && currentMinutes < endMinutes) {
      // We're in the after-midnight portion, and the window started yesterday
      const unblockAt = new Date(t)
      unblockAt.setHours(endParsed.hour, endParsed.minute, 0, 0)
      return { blocked: true, unblockAt }
    }
  } else {
    // Same-day window
    if (containsDay(window.days, dayName) && currentMinutes >= startMinutes && currentMinutes < endMinutes) {
      const unblockAt = new Date(t)
      unblockAt.setHours(endParsed.hour, endParsed.minute, 0, 0)
      return { blocked: true, unblockAt }
    }
  }

  return { blocked: false, unblockAt: null }
}

export interface OffHoursStatus {
  blocked: boolean
  unblockAt: Date | null
  message: string
}

export const evaluateOffHours = (t: Date, config: OffHoursConfig): OffHoursStatus => {
  if (!config.enabled || !config.windows || config.windows.length === 0) {
    return { blocked: false, unblockAt: null, message: '' }
  }

  let earliestUnblock: Date | null = null

  for (const window of config.windows) {
    const result = isTimeInWindow(t, window)
    if (result.blocked) {
      if (result.unblockAt !== null) {
        if (earliestUnblock === null || result.unblockAt < earliestUnblock) {
          earliestUnblock = result.unblockAt
        }
      }
    }
  }

  if (earliestUnblock !== null) {
    return {
      blocked: true,
      unblockAt: earliestUnblock,
      message: config.message || DEFAULT_MESSAGE,
    }
  }

  return { blocked: false, unblockAt: null, message: '' }
}

export const isBlockedNow = async (): Promise<OffHoursStatus> => {
  const config = await fetchOffHoursConfig()
  return evaluateOffHours(new Date(), config)
}

export const getOffHoursMessage = (config: OffHoursConfig): string => {
  return config.message || DEFAULT_MESSAGE
}

File: common/off_hours.go
package common

import (
	"fmt"
	"strings"
	"time"
)

// OffHoursConfig configures optional time-based blocking of sidekick usage.
// A slightly humorous name for when you should be sleeping instead.
type OffHoursConfig struct {
	// Message to display when blocked. Defaults to "Time to rest!" if empty.
	Message string `koanf:"message,omitempty"`
	// Windows defines the time windows during which sidekick is blocked.
	Windows []OffHoursWindow `koanf:"windows,omitempty"`
}

// OffHoursWindow defines a recurring time window for blocking.
type OffHoursWindow struct {
	// Days specifies which days of the week this window applies to.
	// Use lowercase day names: "monday", "tuesday", etc.
	// If empty, applies to all days.
	Days []string `koanf:"days,omitempty" json:"days,omitempty"`
	// Start is the start time in "HH:MM" 24-hour format (local time).
	Start string `koanf:"start" json:"start"`
	// End is the end time in "HH:MM" 24-hour format (local time).
	// If End < Start, the window crosses midnight.
	End string `koanf:"end" json:"end"`
}

// OffHoursStatus represents the current blocking status.
type OffHoursStatus struct {
	Blocked   bool       `json:"blocked"`
	UnblockAt *time.Time `json:"unblock_at,omitempty"`
	Message   string     `json:"message,omitempty"`
}

var dayNameToWeekday = map[string]time.Weekday{
	"sunday":    time.Sunday,
	"monday":    time.Monday,
	"tuesday":   time.Tuesday,
	"wednesday": time.Wednesday,
	"thursday":  time.Thursday,
	"friday":    time.Friday,
	"saturday":  time.Saturday,
}

// IsOffHoursBlockedAt checks if the given time falls within any configured off-hours window.
func IsOffHoursBlockedAt(t time.Time, cfg OffHoursConfig) OffHoursStatus {
	if len(cfg.Windows) == 0 {
		return OffHoursStatus{Blocked: false}
	}

	message := cfg.Message
	if message == "" {
		message = "Time to rest!"
	}

	for _, window := range cfg.Windows {
		blocked, unblockAt := isTimeInWindow(t, window)
		if blocked {
			return OffHoursStatus{
				Blocked:   true,
				UnblockAt: unblockAt,
				Message:   message,
			}
		}
	}

	return OffHoursStatus{Blocked: false}
}

// parseTimeOfDay parses a "HH:MM" string into hour and minute.
func parseTimeOfDay(s string) (hour, minute int, err error) {
	parts := strings.Split(s, ":")
	if len(parts) != 2 {
		return 0, 0, fmt.Errorf("invalid time format: %s", s)
	}
	_, err = fmt.Sscanf(s, "%d:%d", &hour, &minute)
	if err != nil {
		return 0, 0, fmt.Errorf("invalid time format: %s", s)
	}
	if hour < 0 || hour > 23 || minute < 0 || minute > 59 {
		return 0, 0, fmt.Errorf("invalid time values: %s", s)
	}
	return hour, minute, nil
}

// timeOfDayMinutes converts hour:minute to minutes since midnight.
func timeOfDayMinutes(hour, minute int) int {
	return hour*60 + minute
}

// isTimeInWindow checks if time t falls within the given window.
// Returns whether blocked and when the block ends.
func isTimeInWindow(t time.Time, window OffHoursWindow) (blocked bool, unblockAt *time.Time) {
	startH, startM, err1 := parseTimeOfDay(window.Start)
	endH, endM, err2 := parseTimeOfDay(window.End)
	if err1 != nil || err2 != nil {
		return false, nil
	}

	startMins := timeOfDayMinutes(startH, startM)
	endMins := timeOfDayMinutes(endH, endM)
	currentMins := timeOfDayMinutes(t.Hour(), t.Minute())
	isOvernightWindow := endMins < startMins

	// Check day restrictions
	if len(window.Days) > 0 {
		currentDay := strings.ToLower(t.Weekday().String())
		yesterdayName := strings.ToLower(t.AddDate(0, 0, -1).Weekday().String())

		currentDayMatches := containsDay(window.Days, currentDay)
		yesterdayMatches := containsDay(window.Days, yesterdayName)

		if isOvernightWindow {
			// For overnight windows with day restrictions:
			// - Evening portion (after start): check if current day matches
			// - Morning portion (before end): check if yesterday matches
			inMorningPortion := currentMins < endMins
			inEveningPortion := currentMins >= startMins

			if inMorningPortion && yesterdayMatches {
				unblock := time.Date(t.Year(), t.Month(), t.Day(), endH, endM, 0, 0, t.Location())
				return true, &unblock
			}
			if inEveningPortion && currentDayMatches {
				unblock := time.Date(t.Year(), t.Month(), t.Day()+1, endH, endM, 0, 0, t.Location())
				return true, &unblock
			}
			return false, nil
		}

		// Same-day window: just check current day
		if !currentDayMatches {
			return false, nil
		}
	}

	// No day restrictions or same-day window with matching day
	if !isOvernightWindow {
		// Same-day window (e.g., 09:00 to 17:00)
		if currentMins >= startMins && currentMins < endMins {
			unblock := time.Date(t.Year(), t.Month(), t.Day(), endH, endM, 0, 0, t.Location())
			return true, &unblock
		}
	} else {
		// Overnight window without day restrictions (e.g., 23:00 to 07:00)
		if currentMins >= startMins {
			unblock := time.Date(t.Year(), t.Month(), t.Day()+1, endH, endM, 0, 0, t.Location())
			return true, &unblock
		} else if currentMins < endMins {
			unblock := time.Date(t.Year(), t.Month(), t.Day(), endH, endM, 0, 0, t.Location())
			return true, &unblock
		}
	}

	return false, nil
}

// containsDay checks if the days slice contains the given day (case-insensitive).
func containsDay(days []string, day string) bool {
	for _, d := range days {
		if strings.ToLower(d) == day {
			return true
		}
	}
	return false
}
